src/System.Runtime/src/System/HashCode.cs

```diff
@@ -0,0 +1,250 @@
+                    {
+                        queue[position] = val;
+                    }
+                    else
+                    {
+
+                        var sentinel =
+                            state[0] | state[1] | state[2] | state[3] | // any of accumulators set?
+                            (state[Length] ^ position); // length greater than 3?
+
+                        if (sentinel == 0)
+                        {
+                            // Initialize
+
+                            state[0] = Prime1 + Prime2;
+                            state[1] = Prime2;
+                            state[2] = 0;
+                            state[3] = (uint)-Prime1;
+                        }
+
+                        for (var i = 0; i < 3; i++)
+                            state[i] = Round(state[i], queue[i]);
+                        state[3] = Round(state[3], val);
+                    }
+
+                    state[Length]++;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [TargetedPatchingOptOut("Performance critical for inlining across NGen images.")]
+        public int ToHashCode()
+        {
+            unchecked
+            {
+                fixed (uint* state = _state)
+                {
+                    var queue = state + Queue;
+                    var position = state[Length] & 0x3;
+
+                    // Mix the accumulators (if they contain anything)
+
+                    var hash = state[Length] > 3
+                        ? Rol(state[0], 1) + Rol(state[1], 7) + Rol(state[2], 12) + Rol(state[3], 18)
+                        : Prime5;
+
+                    hash += state[Length] * 4;
+
+                    // Mix what remains in the queue
+
+                    for (var i = 0; i < position; i++)
+                    {
+                        hash += queue[i] * Prime3;
+                        hash = Rol(hash, 17) * Prime4;
+                    }
+
+                    // Final mix
+
+                    hash ^= hash >> 15;
+                    hash *= Prime2;
+                    hash ^= hash >> 13;
+                    hash *= Prime3;
+                    hash ^= hash >> 16;
+
+                    return (int)hash;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [TargetedPatchingOptOut("Performance critical for inlining across NGen images.")]
+        private uint Round(uint seed, uint input)
+        {
+            unchecked
+            {
+                seed += input * Prime2;
+                seed = Rol(seed, 13);
+                seed *= Prime1;
+                return seed;
+            }
+        }
+
+#       pragma warning disable 0809
+        [Obsolete("Use ToHashCode to retrieve the computed hash code.", error: true)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override int GetHashCode() => ToHashCode();
```

Comment:
My (and also @tannergooding 's) interpretation of the proposed API was to `throw new NotImplementedException();` here. Usually, and also for this implementation, `ToHashCode()` destroys state, i.e. calling `.ToHashCode()` twice will return different results. For `.GetHashCode()` this would be more than just unexpected, it would lead to wrong results for all the usual consumers (Dictionary etc.). Failing with `InvalidOperationException` might be even better.

Reply:
Happy to do this, just need a final say on which Exception and what message.